cmd/kubeadm/app/cmd/certs.go

```diff
@@ -418,6 +418,10 @@ func fetchCertificateExpirationInfo(rm *renewal.Manager) (*outputapiv1alpha3.Cer
 	return info, nil
 }
 
+// clientSetFromFile is a variable that holds the function to create a clientset from a kubeconfig file.
+// It is used for testing purposes.
+var clientSetFromFile = kubeconfigutil.ClientSetFromFile
```

Comment:
LGTM for test purpose.

Reply:
would regular users still get the extra output when they call the -o json?

i don't think this is the right fix.
see my previous comment on the PR.

Reply:
> if we cannot fix that let's remove the printing from the fetchConfig... function in a separate commit in this PR.
we can also use klog.V(2) as klog goes to stderr.

I think we have to take this if so.

Reply:
https://github.com/kubernetes/kubernetes/pull/108447 introduced many `printer.Print` to replace `fmt.Print` and this will make a big problem to do `kubeadm xx -o json | jq .`.

In another word, we should replace all of them with klog then, IIUC.

Reply:
> If use klog, -o json will still print it. But, certs check-expiration -o json | jq . will work as klog messages will not be handled by jq then.

yes, klog goes on stderr

> printer.Printf("[%s] Reading configuration from the cluster...\n", logPrefix) 
 printer.Printf("[%s] FYI: You can look at this config file with 'kubectl -n %s get cm %s -o yaml'\n", logPrefix, metav1.NamespaceSystem, constants.KubeadmConfigConfigMap) 
  

why is this not working BTW?
we used to be able to have this type of separate - Printf is visible only in -o text and not in -o json.

instead of using klog we should try to understand what is causing the above bug.
if we cannot figure it out before 1.30 code freeze March 6th we can use klog, but that's less preferred.

Reply:
> #108447 introduced many `printer.Print` to replace `fmt.Print` and this will make a big problem to do `kubeadm xx -o json | jq .`.
> 
> In another word, we should replace all of them with klog then, IIUC.

that's way we need to understand why -o json doesn't work properly.
we had a similar problem for `kubeadm config images list` but there -o json works fine AFAIK
can you please test?

if that works, then "plan" and "check-expiration" are doing something wrong.

Reply:
> why is this not working BTW?

I should with after https://github.com/kubernetes/kubernetes/pull/123372/files#r1497230978. 

> would regular users still get the extra output when they call the -o json?

I think regular user with `-o json` should not get the extra output.